hedera-node/src/main/java/com/hedera/services/state/tasks/SystemTaskManager.java

```diff
@@ -0,0 +1,111 @@
+import org.apache.logging.log4j.Logger;
+
+@Singleton
+public class SystemTaskManager {
+    private static final Logger log = LogManager.getLogger(SystemTaskManager.class);
+    private final SystemTask[] tasks;
+
+    @Inject
+    public SystemTaskManager(final Map<String, SystemTask> namedTasks) {
+        // Only runs once per restart, so performance doesn't matter
+        tasks =
+                namedTasks.keySet().stream()
+                        .sorted()
+                        .map(namedTasks::get)
+                        .toArray(SystemTask[]::new);
+        log.info("Discovered {} tasks ({})", tasks.length, namedTasks.keySet());
+    }
+
+    /**
+     * Performs as many pending system tasks as possible for the given entity number and consensus
+     * time. These tasks will begin from {@link MerkleNetworkContext#nextTaskTodo()} and continue
+     * until either the manager receives {@link SystemTaskResult#NO_CAPACITY_LEFT} or {@link
+     * SystemTaskResult#NEEDS_DIFFERENT_CONTEXT}; or until the end of the {@code tasks} list is
+     * reached.
+     *
+     * <p>More precisely, this method will return:
+     *
+     * <ul>
+     *   <li>{@link SystemTaskResult#NOTHING_TO_DO} iff all tasks returned {@code NOTHING_TO_DO}.
+     *   <li>{@link SystemTaskResult#DONE} iff at least one task returned {@code DONE} and all other
+     *       tasks returned {@code NOTHING_TO_DO}.
+     *   <li>{@link SystemTaskResult#NO_CAPACITY_LEFT} immediately when any task reports it.
+     *   <li>{@link SystemTaskResult#NEEDS_DIFFERENT_CONTEXT} immediately when any task reports it.
+     * </ul>
+     *
+     * @param literalNum the id to process
+     * @param now the consensus time now
+     * @param curNetworkCtx the current network context
+     * @return the cumulative result of attempting all pending tasks
+     */
+    public SystemTaskResult process(
+            final long literalNum, final Instant now, final MerkleNetworkContext curNetworkCtx) {
+        boolean didSomething = false;
+        for (var i = curNetworkCtx.nextTaskTodo(); i < tasks.length; i++) {
+            if (!tasks[i].isActive(literalNum, curNetworkCtx)) {
+                continue;
+            }
+            final var result = tasks[i].process(literalNum, now);
+            if (result == NOTHING_TO_DO || result == DONE) {
+                didSomething |= (result == DONE);
+            } else if (result == NEEDS_DIFFERENT_CONTEXT || result == NO_CAPACITY_LEFT) {
+                curNetworkCtx.setNextTaskTodo(i);
+                return result;
+            }
+        }
+        curNetworkCtx.setNextTaskTodo(0);
+        // Only update the pre-existing scan status when we complete all tasks for an entity id
+        updatePreExistingScanStatus(literalNum, curNetworkCtx);
+        return didSomething ? DONE : NOTHING_TO_DO;
+    }
+
+    @VisibleForTesting
+    void updatePreExistingScanStatus(final long n, final MerkleNetworkContext curNetworkCtx) {
```

Comment:
I thought sonar will warn if we have any method being used in production code with `@VisibleForTesting` 🤔

Reply:
Sonar is down right now, you're probably right 😁